apps/studio-next/package.json

```diff
@@ -6,7 +6,14 @@
     "dev": "next dev -p 3001",
     "build": "next build",
     "start": "next start",
-    "lint": "echo 'No linting configured'"
+    "lint": "next lint",
+    "docker:build": "docker build -t asyncapi/studio:latest .",
```

Comment:
docker build should be done at the root of the project (so turborepo is able to prune the unwanted packages). maybe something like: `"build:next-studio:docker": "docker build -t asyncapi/studio:latest -f StudioNextDockerfile ."` at the root.

Reply:
Simply making sure that Turborepo will not start to be turned into a monolith again after this change.

Might it be better, just in case, to keep possibility to dockerize each app separately with [Docker Compose](https://docs.docker.com/compose/intro/features-uses), like it's done in https://github.com/dhaaana/turborepo-with-docker:

- https://github.com/dhaaana/turborepo-with-docker/blob/main/docker-compose.yml
- https://github.com/dhaaana/turborepo-with-docker/blob/main/apps/api/Dockerfile
- https://github.com/dhaaana/turborepo-with-docker/blob/main/apps/web/Dockerfile

?
(meaning `Dockerfile` will stay where it is now, but `Docker Compose` setup will be added)

Reply:
If it's possible to [publish the separate docker image](https://github.com/asyncapi/studio/blob/master/.github/workflows/publish-docker-image.yml) after this, I have no issue with it.

Reply:
`turbo prune studio-next --docker` copies to `./out`'s `node_modules` containing `pnpm`'s symlinks, which, after being copied, immediately become invalid (because they point to previous relative destinations,) but doing `pnpm install` each time there's a need to build a Docker image is not an option either. So I'm trying different configurations to achieve local caching in at least some form, and each build takes time. Thus delay.

Reply:
I changed approach in favor of pushing into the Docker image the **already compiled** `Studio`.
I took advantage of `turbo build` cache, and now if there is a fresh `turbo`'s build cache, then the generation of the Docker image takes not even seconds but milliseconds.

![image](https://github.com/asyncapi/studio/assets/16149591/38bf0837-6c33-4d9f-898d-291acb373bf6)

Please try running
```
turbo build
turbo run build:studio-next:docker
```
in the root of the repo and check.

The generated Docker image (became 294MB) is run with
```
docker run -p 3001:3001 asyncapi/studio
```

Reply:
> turbo prune studio-next --docker copies to ./out's node_modules containing pnpm's symlinks, which, after being copied, immediately become invalid

That's why we use ` turbo prune` to get rid of the unwanted packages before running `pnpm install`. after we have made sure to separate the wanted from the unwanted packages.

after that we install the packages inside of the docker container.

> I changed approach in favor of pushing into the Docker image the already compiled Studio.

well, we will run the docker build command on the GitHub Actions workflow. so there will not be any *already compiled* studio to build the docker image from.

We would still have to install dependencies for the entire mono-repo, re-build the entire project, then use the built project to build the docker image.

even if it's easier for us to build, think about the user of the image. if we push one big un-cacheable layer (like we currently do) every time something changes (even if it doesn't affect studio-next), it will make the build command for them to take soooo much longer.

why not follow the [official approach](https://turbo.build/repo/docs/guides/tools/docker#example) which would build the studio on layers and then below layers will be cached anyway for the end user, so they will only download what they need( for example if no new dependency added, they will not install the dependencies again, they will get the source and continue from there).

At the end, it doesn't matter what the final size of the image is, docker won't install all of it, it will cache the steps that are possible to cache and only builds from the layer that the change is detected. we can help docker by separating our steps(first, copy package.json, then install dependencies, then copy the source code, then build, etc...) so it can properly cache each step.

Reply:
btw, I am not an expert in docker. please let me know if I misunderstand something here. 😆

Reply:
`turbo prune studio-next --docker` creates in `./out/full/apps/studio-next/node_modules` set of symlinks to all used Node.js' modules, which are invalid even on local filesystem, let alone in the Docker image filesystem after they are pushed to it.

![image](https://github.com/asyncapi/studio/assets/16149591/e0121305-1598-4bf5-a48f-7d804e7d665f)
![image](https://github.com/asyncapi/studio/assets/16149591/6f5aec7e-e30c-48c1-a75b-e40999644f20)

Is it different for you?

To my understanding, `turbo prune studio-next --docker` would be useful if `Studio` used `npm`, but in the case of `pnpm`, it's useless exactly due to symlinking of Node.js modules. That's why, after trying to get it working for a week, at the end of the day, I changed approach to pushing to the Docker image the already locally built binaries.

Might I simply be doing something wrong?


```
"build:studio-next:docker": {
  "dependsOn": ["^build:studio-next:docker"]
}
```
is hardcoded to run `next build` as the first step
https://github.com/asyncapi/studio/pull/1094/files#diff-c493037e1471391d87ade822ee79d21c8ba48c14063c5feafdf629443b19cc0eR10

So either there is already a fresh build and the `turbo`'s cache will be utilized, or `Studio` will be built from scratch anyway.

Aside from that, I also think that release process can have two simultaneous deployments:
- Build and deploy to the **studio.asyncapi.com**.
- Reuse already built binaries to generate a Docker image and deploy it to the **Docker hub**.

Such a release process will save time and space, and it seems quite logical overall.

Reply:
@aeworxet 
Why are we copying the `node_modules` folder to the docker?
I think we are supposed to ignore it and install it fresh in the docker environment as a separate step.
This way it will be cached by docker and we never have to reinstall the packages unless a package.json file changes.

can you try adding `node_modules` to the `.dockerignore` file?

Reply:
I put together this one which I think you can use to improve upon:
btw you need to enable docker buildKit for pnpm caching to work. not sure how we can enable it in Github Actions.
```dockerfile
FROM node:18-alpine AS base


FROM base AS builder
RUN apk update
RUN apk add --no-cache libc6-compat

ENV PNPM_HOME="/pnpm"
ENV PATH="$PNPM_HOME:$PATH"
RUN corepack enable

WORKDIR /app
RUN pnpm add -g turbo@^1

COPY . .
ENV NODE_ENV production
RUN turbo prune studio-next --docker

FROM base AS installer
ENV DOCKER_BUILDKIT=1
RUN apk update
RUN apk add --no-cache libc6-compat

ENV PNPM_HOME="/pnpm"
ENV PATH="$PNPM_HOME:$PATH"
RUN corepack enable
WORKDIR /app

COPY .gitignore .gitignore
COPY --from=builder /app/out/json/ .
COPY --from=builder /app/out/pnpm-lock.yaml ./pnpm-lock.yaml

RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install --ignore-scripts --frozen-lockfile

COPY --from=builder /app/out/full/ .
RUN pnpm run build:studio-next

FROM base AS runner

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs
USER nextjs

COPY --from=installer /app/apps/studio-next/next.config.js .
COPY --from=installer /app/apps/studio-next/package.json .

COPY --from=installer --chown=nextjs:nodejs /app/apps/studio-next/.next/standalone ./
COPY --from=installer --chown=nextjs:nodejs /app/apps/studio-next/.next/static ./apps/studio-next/.next/static
COPY --from=installer --chown=nextjs:nodejs /app/apps/studio-next/public ./apps/studio-next/public


CMD node apps/studio-next/server.js

```

Reply:
`node_modules` [were ignored initially](https://github.com/asyncapi/studio/pull/1094/files#diff-2f754321d62f08ba8392b9b168b83e24ea2852bb5d815d63e767f6c3d23c6ac5L3).
`pnpm` was installing them from scratch inside the Docker image.
[You said](https://github.com/asyncapi/studio/pull/1094#discussion_r1648838607) Installing them takes too long.
I changed approach to pushing to the Docker image the pre-compiled as `standalone` Next.js app.
`node_modules` are an integral part of the `standalone` version (`export` version is fully static.)
Docker didn't copy `node_modules` to the image because they were in `.dockerignore`, making `standalone` Next.js app unusable.
I had to remove the `node_modules` from `.dockerignore`.

It is already optimized during compilation of the Next.js app for production, though, and takes only 152 MB of code, not 1.9 G of root `node_modules`.

Reply:
@derberg
Can you please suggest a Docker expert?
I'm afraid we're starting to go in circles here.

Reply:
@aeworxet 

> I had to remove the node_modules from .dockerignore.

I never suggested that you remove the `node_modules` from docker container or change your approach. all I said was to add the `--ignore-scripts` flag so the browser is not downloaded by Cypress during installation.

> You said Installing them takes too long.

Yes, it takes too long. but one part of the solution is to not run the post-install scripts which [I suggested here](https://github.com/asyncapi/studio/pull/1094#discussion_r1648838607) and the second part would be to make the pnpm use it's cached packages which I [suggested here](https://github.com/asyncapi/studio/pull/1094#discussion_r1668359299).

> I changed approach to pushing to the Docker image the pre-compiled as standalone Next.js app.

Well, the problem with that approach is that docker can't cache the result of the build steps and the user of the docker image has to pull the entire image for each update. which [I pointed out here](https://github.com/asyncapi/studio/pull/1094#discussion_r1668291253).

> It is already optimized during compilation of the Next.js app for production, though, and takes only 152 MB of code, not 1.9 G of root node_modules.

What do you think the image size would be if you follow [the official approach](https://turbo.build/repo/docs/guides/tools/docker)? Which I suggested twice ([here](https://github.com/asyncapi/studio/pull/1094#discussion_r1668359299), and [here](https://github.com/asyncapi/studio/pull/1094#discussion_r1648834593)) by now in this thread.

Let's take a look at the sizes with the two approaches:
<img width="985" alt="Screenshot 2024-07-08 at 14 05 36" src="https://github.com/asyncapi/studio/assets/32505158/9ebfc54f-7bf0-45f3-ac3d-2916c41c7c41">
As you can see the official approach's image size is even smaller than what [you claim](https://github.com/asyncapi/studio/pull/1094#discussion_r1662117391) your approaches' size is.
(Maybe something is wrong here since your image size comes up as 3.19GB on my machine.)

> I'm afraid we're starting to go in circles here.

I don't exactly get what you mean.
1) You implemented a solution
2) the solution had a problem(pnpm packages took too long to install).
3) I pointed out the problem and suggested [a possible solution](https://github.com/asyncapi/studio/pull/1094#discussion_r1648838607)
4) you changed the ENTIRE approach.
5) I pointed out the [problems with the new approach](https://github.com/asyncapi/studio/pull/1094#discussion_r1668291253) and suggested [a solution](https://github.com/asyncapi/studio/pull/1094#discussion_r1668359299) which resolves the problem. (by not running post-install scripts and using pnpm's cache)
6) you claim we are going full circle and need someone else to review your work?

Reply:
I clearly [misperformed](https://github.com/asyncapi/studio/issues/661#issuecomment-2215396594) on this Bounty Issue.
Should Docker-related code be removed so other changes could be merged?

Reply:
Just built the Docker image with `turbo build:studio-next:docker`

![image](https://github.com/asyncapi/studio/assets/16149591/c337c85e-ff07-4070-afb9-a11d290fa03e)

The difference is strange.

Reply:
Maybe you should clone this branch from scratch and do in the repo's root
```bash
pnpm install
docker builder prune -a    # <-- remove Docker's build cache
turbo build:studio-next:docker
```
?

Just built again with these steps

![image](https://github.com/asyncapi/studio/assets/16149591/6c889eae-a8a2-4f91-b5e4-7c2b7a95157d)

Reply:
Upon some reflection I think my tone might have come across as too critical and for that I apologize.

My intentions were to provide technical feedback not to criticize you or your efforts personally.

I know you have worked hard on the issue and I don't think it was just for you to be suspended in the bounty program or not getting the reward for this bounty issue.

If you think it's best to remove the docker and merge the remaining code, let's do that.

Reply:
> Maybe you should clone this branch from scratch and do in the repo's root
> ```bash
> pnpm install
> docker builder prune -a    # <-- remove Docker's build cache
> turbo build:studio-next:docker
> ```
> ?
> 
> Just built again with these steps
> 
> ![image](https://github.com/asyncapi/studio/assets/16149591/6c889eae-a8a2-4f91-b5e4-7c2b7a95157d)
> 

I cleaned the project reinstalled the packages and built the image with `turbo build:studio-next:docker`

The size now shows up correctly.

Reply:
What do you think about reusing the same compiled code for two simultaneous deployments now?

> release process can have two simultaneous deployments:
> 
> - Build and deploy to the **studio.asyncapi.com**.
> - Reuse already built binaries (with `turbo` cache) to generate (push them into) a Docker image and deploy it to the **Docker hub**.

This process will guarantee that there is one same `Studio`, built from one same commit, with the same hashes, in both places.
(I already have a situation right now when the same bug gets reproduced in one place, doesn't in another, and gives a third error in the local dev)

Reply:
@aeworxet

### Building in One Environment and Running in Another Is Generally a Bad Idea
As you know, we are using linux to run random workflows. However, when it comes to publishing or testing, we are using a matrix of (Linux, Windows, and Mac) to [build, publish or test](https://github.com/asyncapi/cli/blob/master/.github/workflows/if-nodejs-release.yml) our code.

Why? Because we want to make sure that it is built within the environment that it is going to run on. How can you be sure that if you build something on Linux, it will run on Windows? You might argue that they are different since docker has it's own Engine on top of the OS and it will work. You might be right, but we need to test that theory. Is it worth it to go though the trouble, I don't think so.
All I know is that some packages are platform-dependent. By installing in the Docker image itself, we make sure that all our packages are installed, built, and run in the same environment.

That's why building in the Docker image itself is always better.

### Transparency
When I look at the Dockerfile, I should be able to see that yes, I am actually downloading the source code and building it myself on my machine. This gives me peace of mind that nobody is pushing some fishy code into my environment. Most of the ecosystem now doesn't trust pre-built packages because of incidents like the [XZ fiasco](https://www.youtube.com/watch?v=bS9em7Bg0iU).

### Docker Cache
As I mentioned previously, Docker is really smart when it comes to building and running your image. If you follow best practices, it can help your package be as slim as possible.

For example, if you haven't changed your package.json or lock file, why would you want to install your packages from scratch? Well, you don't want that. You want to get the source code, which is probably a few MB, use the cached version of the package installation layer, and build on top of that, right? Well, Docker is going to do exactly that. But if you are pushing an already built 290MB of code in a single layer, Docker won't be able to do anything. Even if the next version only changes by one byte, it has to invalidate that layer and install the entire 290MB of code again.

So in a Dockerfile, it is always better to break down your build, installation, and run layers as much as you can. More layers mean Docker is able to cache more efficiently when `building` and when `running`.

### Harder to Maintain
It is hard to maintain a separate and custom workflow to do things. People expect simplicity. If I want to build an image, I would naturally expect the `docker build` command to work. why does it have to be dependent on other processes like installation and building?

If you build one more layer of abstraction on top and use a script, it is suspicious to say the least. Contributors/Users will have to check what exactly you are doing in that script and why they can't just build it with a simple `docker build` command.

Reply:
Thanks for investing time in explaining all that.
I now understand that I should have taken specific end users' needs into account also, not only aim at easing the release process for the vendor (AsyncAPI.) 
The Docker build should definitely be developed by someone who understands Docker better.
Removed Docker-related code and changed the PR's description accordingly.

Reply:
Thanks for investing time in explaining all that.
I now understand that I should have taken specific end users' needs into account also, not only aim at easing the release process for the 'vendor.' 
The Docker build should definitely be developed by someone who understands Docker better.
Removed Docker-related code and changed the PR's description accordingly.